File: app/src/main/java/com/bytedance/tiktok/fragment/PersonalHomeFragment.java
package com.bytedance.tiktok.fragment;

import android.content.Intent;
import android.view.View;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.RelativeLayout;
import android.widget.TextView;
import androidx.appcompat.widget.Toolbar;
import androidx.coordinatorlayout.widget.CoordinatorLayout;
import androidx.core.app.ActivityCompat;
import androidx.core.app.ActivityOptionsCompat;
import androidx.fragment.app.Fragment;
import androidx.viewpager.widget.ViewPager;
import com.androidkun.xtablayout.XTabLayout;
import com.bytedance.tiktok.R;
import com.bytedance.tiktok.activity.FocusActivity;
import com.bytedance.tiktok.activity.ShowImageActivity;
import com.bytedance.tiktok.base.BaseFragment;
import com.bytedance.tiktok.base.CommPagerAdapter;
import com.bytedance.tiktok.bean.CurUserBean;
import com.bytedance.tiktok.bean.MainPageChangeEvent;
import com.bytedance.tiktok.utils.NumUtils;
import com.bytedance.tiktok.utils.RxBus;
import com.bytedance.tiktok.view.CircleImageView;
import com.bytedance.tiktok.view.IconFontTextView;
import com.google.android.material.appbar.AppBarLayout;
import java.util.ArrayList;
import butterknife.BindView;
import rx.functions.Action1;

/**
 * create by libo
 * create on 2020-05-19
 * description 个人主页fragment
 */
public class PersonalHomeFragment extends BaseFragment implements View.OnClickListener {
    @BindView(R.id.iv_bg)
    ImageView ivBg;
    @BindView(R.id.iv_head)
    CircleImageView ivHead;
    @BindView(R.id.rl_dropdown)
    RelativeLayout rlDropdown;
    @BindView(R.id.ll_focus)
    LinearLayout llFocus;
    @BindView(R.id.ll_fans)
    LinearLayout llFans;
    @BindView(R.id.iv_return)
    ImageView ivReturn;
    @BindView(R.id.tv_title)
    TextView tvTitle;
    @BindView(R.id.tv_focus)
    TextView tvFocus;
    @BindView(R.id.iv_more)
    IconFontTextView ivMore;
    @BindView(R.id.toolbar)
    Toolbar toolbar;
    @BindView(R.id.tablayout)
    XTabLayout tabLayout;
    @BindView(R.id.appbarlayout)
    AppBarLayout appBarLayout;
    @BindView(R.id.viewpager)
    ViewPager viewPager;
    @BindView(R.id.tv_nickname)
    TextView tvNickname;
    @BindView(R.id.tv_sign)
    TextView tvSign;
    @BindView(R.id.tv_getlike_count)
    TextView tvGetLikeCount;
    @BindView(R.id.tv_focus_count)
    TextView tvFocusCount;
    @BindView(R.id.tv_fans_count)
    TextView tvFansCount;
    private ArrayList<Fragment> fragments = new ArrayList<>();
    private CommPagerAdapter pagerAdapter;
    String[] titleStr = new String[]{"作品 " , "动态 " , "喜欢 " };


    @Override
    protected int setLayoutId() {
        return R.layout.fragment_personal_home;
    }

    @Override
    protected void init() {

        //解决toolbar左边距问题
        toolbar.setContentInsetsAbsolute(0, 0);

        setAppbarLayoutPercent();

        ivReturn.setOnClickListener(this);
        ivHead.setOnClickListener(this);
        ivBg.setOnClickListener(this);
        llFocus.setOnClickListener(this);
        llFans.setOnClickListener(this);

        setUserInfo();

        setTabLayout();
    }

    /**
     * 过渡动画跳转页面
     *
     * @param view
     */
    public void transitionAnim(View view) {
        ActivityOptionsCompat compat = ActivityOptionsCompat.makeSceneTransitionAnimation(getActivity(), view, getString(R.string.trans));
        ActivityCompat.startActivity(getActivity(), new Intent(getActivity(), ShowImageActivity.class), compat.toBundle());
    }

    public void setUserInfo() {
        RxBus.getDefault().toObservable(CurUserBean.class).subscribe((Action1<CurUserBean>) curUserBean -> {

            coordinatorLayoutBackTop();

            ivBg.setImageResource(curUserBean.getUserBean().getHead());
            ivHead.setImageResource(curUserBean.getUserBean().getHead());
            tvNickname.setText(curUserBean.getUserBean().getNickName());
            tvSign.setText(curUserBean.getUserBean().getSign());
            tvTitle.setText(curUserBean.getUserBean().getNickName());

            String subCount = NumUtils.numberFilter(curUserBean.getUserBean().getSubCount());
            String focusCount = NumUtils.numberFilter(curUserBean.getUserBean().getFocusCount());
            String fansCount = NumUtils.numberFilter(curUserBean.getUserBean().getFansCount());

            tvGetLikeCount.setText(subCount);
            tvFocusCount.setText(focusCount);
            tvFansCount.setText(fansCount);
        });
    }

    private void setTabLayout() {
//        String[] titles = new String[]{"作品 " + userInfo.getWorkCount(), "动态 " + userInfo.getDynamicCount(), "喜欢 " + userInfo.getLikeCount()};

        for (int i = 0; i < titleStr.length; i++) {
            fragments.add(new WorkFragment());
            tabLayout.addTab(tabLayout.newTab().setText(titleStr[i]));
        }

        pagerAdapter = new CommPagerAdapter(getChildFragmentManager(), fragments, titleStr);
        viewPager.setAdapter(pagerAdapter);
        tabLayout.setupWithViewPager(viewPager);
    }

    /**
     * 根据滚动比例渐变view
     */
    private void setAppbarLayoutPercent() {
        appBarLayout.addOnOffsetChangedListener((appBarLayout, verticalOffset) -> {
            float percent = (Math.abs(verticalOffset * 1.0f) / appBarLayout.getTotalScrollRange());  //滑动比例

            if (percent > 0.8) {
                tvTitle.setVisibility(View.VISIBLE);
                tvFocus.setVisibility(View.VISIBLE);

                float alpha = 1 - (1 - percent) * 5;  //渐变变换
                tvTitle.setAlpha(alpha);
                tvFocus.setAlpha(alpha);
            } else {
                tvTitle.setVisibility(View.GONE);
                tvFocus.setVisibility(View.GONE);
            }
        });
    }

    /**
     * 自动回顶部
     */
    private void coordinatorLayoutBackTop() {
        CoordinatorLayout.Behavior behavior =
                ((CoordinatorLayout.LayoutParams) appBarLayout.getLayoutParams()).getBehavior();
        if (behavior instanceof AppBarLayout.Behavior) {
            AppBarLayout.Behavior appBarLayoutBehavior = (AppBarLayout.Behavior) behavior;
            int topAndBottomOffset = appBarLayoutBehavior.getTopAndBottomOffset();
            if (topAndBottomOffset != 0) {
                appBarLayoutBehavior.setTopAndBottomOffset(0);
            }
        }
    }

    @Override
    public void onClick(View v) {
        switch (v.getId()) {
            case R.id.iv_return:
                RxBus.getDefault().post(new MainPageChangeEvent(0));
                break;
            case R.id.iv_head:
                transitionAnim(ivHead);
                break;
            case R.id.iv_bg:
                transitionAnim(ivBg);
                break;
            case R.id.ll_focus:
                startActivity(new Intent(getActivity(), FocusActivity.class));
                break;
            case R.id.ll_fans:
                startActivity(new Intent(getActivity(), FocusActivity.class));
                break;
        }
    }
}

File: app/src/main/java/com/bytedance/tiktok/fragment/PersonalHomeFragment.kt
package com.bytedance.tiktok.fragment

import android.content.Intent
import android.os.Bundle
import android.view.View
import androidx.coordinatorlayout.widget.CoordinatorLayout
import androidx.core.app.ActivityCompat
import androidx.core.app.ActivityOptionsCompat
import androidx.fragment.app.Fragment
import com.bytedance.tiktok.R
import com.bytedance.tiktok.activity.FocusActivity
import com.bytedance.tiktok.activity.ShowImageActivity
import com.bytedance.tiktok.base.BaseBindingFragment
import com.bytedance.tiktok.base.CommPagerAdapter
import com.bytedance.tiktok.bean.CurUserBean
import com.bytedance.tiktok.bean.MainPageChangeEvent
import com.bytedance.tiktok.bean.VideoBean.UserBean
import com.bytedance.tiktok.databinding.FragmentPersonalHomeBinding
import com.bytedance.tiktok.utils.NumUtils
import com.bytedance.tiktok.utils.RxBus
import com.google.android.material.appbar.AppBarLayout
import rx.Subscription
import rx.functions.Action1
import java.util.*

/**
 * create by libo
 * create on 2020-05-19
 * description 个人主页fragment
 */
class PersonalHomeFragment : BaseBindingFragment<FragmentPersonalHomeBinding>({FragmentPersonalHomeBinding.inflate(it)})
    , View.OnClickListener {
    private val fragments = ArrayList<Fragment>()
    private var pagerAdapter: CommPagerAdapter? = null
    private var curUserBean: UserBean? = null
    private var subscription: Subscription? = null

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        init()
    }

    fun init() {

        //解决toolbar左边距问题
        binding.toolbar!!.setContentInsetsAbsolute(0, 0)
        setappbarlayoutPercent()
        binding.ivReturn!!.setOnClickListener(this)
        binding.homeHeader.ivHead!!.setOnClickListener(this)
        binding.ivBg!!.setOnClickListener(this)
        binding.homeHeader.llFocus!!.setOnClickListener(this)
        binding.homeHeader.llFans!!.setOnClickListener(this)
        setUserInfo()
    }

    /**
     * 过渡动画跳转页面
     *
     * @param view
     */
    fun transitionAnim(view: View, res: Int) {
        val compat = ActivityOptionsCompat.makeSceneTransitionAnimation(requireActivity(), view, getString(R.string.trans))
        val intent = Intent(activity, ShowImageActivity::class.java)
        intent.putExtra("res", res)
        ActivityCompat.startActivity(requireActivity(), intent, compat.toBundle())
    }

    fun setUserInfo() {
        subscription = RxBus.getDefault().toObservable(CurUserBean::class.java).subscribe(Action1 { curUserBean: CurUserBean ->
            coordinatorLayoutBackTop()
            this.curUserBean = curUserBean.userBean
            binding.ivBg!!.setImageResource(curUserBean.userBean.head)
            binding.homeHeader.ivHead!!.setImageResource(curUserBean.userBean.head)
            binding.homeHeader.tvNickname!!.text = curUserBean.userBean.nickName
            binding.homeHeader.tvSign!!.text = curUserBean.userBean.sign
            binding.tvTitle!!.text = curUserBean.userBean.nickName
            val subCount = NumUtils.numberFilter(curUserBean.userBean.subCount)
            val focusCount = NumUtils.numberFilter(curUserBean.userBean.focusCount)
            val fansCount = NumUtils.numberFilter(curUserBean.userBean.fansCount)

            //获赞 关注 粉丝
            binding.homeHeader.tvGetLikeCount!!.text = subCount
            binding.homeHeader.tvFocusCount!!.text = focusCount
            binding.homeHeader.tvFansCount!!.text = fansCount

            //关注状态
            if (curUserBean.userBean.isFocused) {
                binding.homeHeader.tvAddfocus!!.text = "已关注"
                binding.homeHeader.tvAddfocus!!.setBackgroundResource(R.drawable.shape_round_halfwhite)
            } else {
                binding.homeHeader.tvAddfocus!!.text = "关注"
                binding.homeHeader.tvAddfocus!!.setBackgroundResource(R.drawable.shape_round_red)
            }
            setTabLayout()
        } as Action1<CurUserBean>)
    }

    private fun setTabLayout() {
        val titles = arrayOf("作品 " + curUserBean!!.workCount, "动态 " + curUserBean!!.dynamicCount, "喜欢 " + curUserBean!!.likeCount)
        fragments.clear()
        for (i in titles.indices) {
            fragments.add(WorkFragment())
            binding.tabLayout!!.addTab(binding.tabLayout!!.newTab().setText(titles[i]))
        }
        pagerAdapter = CommPagerAdapter(childFragmentManager, fragments, titles)
        binding.viewPager!!.adapter = pagerAdapter
        binding.tabLayout!!.setupWithViewPager(binding.viewPager)
    }

    /**
     * 根据滚动比例渐变view
     */
    private fun setappbarlayoutPercent() {
        binding.appbarlayout.addOnOffsetChangedListener(AppBarLayout.OnOffsetChangedListener { appbarlayout, verticalOffset ->
            val percent = Math.abs(verticalOffset * 1.0f) / appbarlayout.totalScrollRange //滑动比例
            if (percent > 0.8) {
                binding.tvTitle!!.visibility = View.VISIBLE
                binding.tvFocus!!.visibility = View.VISIBLE
                val alpha = 1 - (1 - percent) * 5 //渐变变换
                binding.tvTitle!!.alpha = alpha
                binding.tvFocus!!.alpha = alpha
            } else {
                binding.tvTitle!!.visibility = View.GONE
                binding.tvFocus!!.visibility = View.GONE
            }
        })
    }

    /**
     * 自动回顶部
     */
    private fun coordinatorLayoutBackTop() {
        val behavior = (binding.appbarlayout!!.layoutParams as CoordinatorLayout.LayoutParams).behavior
        if (behavior is AppBarLayout.Behavior) {
            val appbarlayoutBehavior = behavior
            val topAndBottomOffset = appbarlayoutBehavior.topAndBottomOffset
            if (topAndBottomOffset != 0) {
                appbarlayoutBehavior.topAndBottomOffset = 0
            }
        }
    }

    override fun onClick(v: View) {
        when (v.id) {
            R.id.ivReturn -> RxBus.getDefault().post(MainPageChangeEvent(0))
            R.id.ivHead -> transitionAnim(binding.homeHeader.ivHead, curUserBean!!.head)
            R.id.ivBg -> {
            }
            R.id.llFocus -> startActivity(Intent(activity, FocusActivity::class.java))
            R.id.llFans -> startActivity(Intent(activity, FocusActivity::class.java))
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        if (subscription != null) {
            subscription!!.unsubscribe()
        }
    }
}